### DISCUSSION TITLE
Can I get ast from preprocessor?

### DISCUSSION THREAD

**Discussion by codehz** [edubart/nelua-lang#104] [Q&A] [answered]
It seems
```lua
## print(
print("hello")
## )
```
will be transform to something like 
`print(inject_astnode(...))`
so I cannot get ast through this way...

I want to transform ast before injecting

Also, is it possible to implement custom annotations?

**Comment by edubart** [comment, marked as answer]
> I want to transform ast before injecting

This functionality is not officially supported yet, but you can "produce" an AST from a block of statements with some meta programming tricks:

```lua
##[[
-- Utility to help getting a ast from a block of statements.
local function produce_ast(f)
  -- lets hook inject_statement to capture the node
  local inject_statement = ppcontext.inject_statement
  local node
  function ppcontext.inject_statement(_, statnode)
    if node then error 'please use multiple statements inside a `do` block' end
    node = statnode
  end
  f()
  ppcontext.inject_statement = inject_statement
  return node
end
]]

## local myast = produce_ast(function()
  print("hello")
## end)

-- Print the AST in table form (for debugging)
## print(myast)

-- Print the AST in compact way (for debugging)
## print(myast:pretty())

## inject_statement(myast)
```

I may make this later an official API, just have to think more on better name and use cases.

> Also, is it possible to implement custom annotations?

Yes, with some modding in the compiler, check this discussion for an example:

https://github.com/edubart/nelua-lang/discussions/70#discussioncomment-832650

Be warned that I may make a cleaner way to add annotations in the future through an official API.
